File: go/src/pkg/container/list/list.go
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
   "http://www.w3.org/TR/html4/transitional.dtd">
<html>
<head>

  <meta http-equiv="content-type" content="text/html; charset=utf-8">
  <title>Source file /src/pkg/container/list/list.go</title>

  <link rel="stylesheet" type="text/css" href="../../../../doc/style.css">
  <script type="text/javascript" src="../../../../doc/godocs.js"></script>

</head>

<body>

  <script>
    // Catch 'enter' key down events and trigger the search form submission.
    function codesearchKeyDown(event) {
      if (event.which == 13) {
        var form = document.getElementById('codesearch');
        var query = document.getElementById('codesearchQuery');
        form.q.value = "lang:go package:go.googlecode.com " + query.value;
        document.getElementById('codesearch').submit();
}      return true;
}
    // Capture the submission event and construct the query parameter.
    function codeSearchSubmit() {
      var query = document.getElementById('codesearchQuery');
      var form = document.getElementById('codesearch');
      form.q.value = "lang:go package:go.googlecode.com " + query.value;
      return true;
}  </script>

<div id="topnav">
  <table summary="">
    <tr>
      <td id="headerImage">
        <a href="../../../../index.html"><img src="../../../../doc/logo-153x55.png" height="55" width="153" alt="Go Home Page" style="border:0" /></a>
      </td>
      <td>
        <div id="headerDocSetTitle">The Go Programming Language</div>
      </td>
      <td>
        <!-- <table>
          <tr>
            <td>
              <! The input box is outside of the form because we want to add
              a couple of restricts to the query before submitting. If we just
              add the restricts to the text box before submitting, then they
              appear in the box when the user presses 'back'. Thus we use a
              hidden field in the form. However, there's no way to stop the
              non-hidden text box from also submitting a value unless we move
              it outside of the form
              <input type="search" id="codesearchQuery" value="" size="30" onkeydown="return codesearchKeyDown(event);"/>
              <form method="GET" action="http://www.google.com/codesearch" id="codesearch" class="search" onsubmit="return codeSearchSubmit();" style="display:inline;">
                <input type="hidden" name="q" value=""/>
                <input type="submit" value="Code search" />
                <span style="color: red">(TODO: remove for now?)</span>
              </form>
            </td>
          </tr>
          <tr>
            <td>
              <span style="color: gray;">(e.g. &ldquo;pem&rdquo; or &ldquo;xml&rdquo;)</span>
            </td>
          </tr>
        </table> -->
      </td>
    </tr>
  </table>
</div>

<div id="linkList">
  <ul>
    <li class="navhead"><a href="../../../../index.html">Home</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Documents</li>
    <li><a href="../../../../doc/go_tutorial.html">Tutorial</a></li>
    <li><a href="../../../../doc/effective_go.html">Effective Go</a></li>
    <li><a href="../../../../doc/go_faq.html">FAQ</a></li>
    <li><a href="../../../../doc/go_lang_faq.html">Language Design FAQ</a></li>
    <li><a href="http://www.youtube.com/watch?v=rKnDgT73v8s">Tech talk (1 hour)</a> (<a href="../../../../doc/go_talk-20091030.pdf">PDF</a>)</li>
    <li><a href="../../../../doc/go_spec.html">Language Specification</a></li>
    <li><a href="../../../../doc/go_mem.html">Memory Model</a></li>
    <li><a href="../../../../doc/go_for_cpp_programmers.html">Go for C++ Programmers</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">How To</li>
    <li><a href="../../../../doc/install.html">Install Go</a></li>
    <li><a href="../../../../doc/contribute.html">Contribute code</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Programming</li>
    <li><a href="../../../../cmd/index.html">Command documentation</a></li>
    <li><a href="../../../../pkg/index.html">Package documentation</a></li>
    <li><a href="../../../index.html">Source files</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Help</li>
    <li>#go-nuts on irc.freenode.net</li>
    <li><a href="http://groups.google.com/group/golang-nuts">Go Nuts mailing list</a></li>
    <li><a href="http://code.google.com/p/go/issues/list">Issue tracker</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Go code search</li>
    <form method="GET" action="http://golang.org/search" class="search">
    <input type="search" name="q" value="" size="25" style="width:80%; max-width:200px" />
    <input type="submit" value="Go" />
    </form>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Last update</li>
	<li>Thu Nov 12 15:48:37 PST 2009</li>
  </ul>
</div>

<div id="content">
  <h1 id="generatedHeader">Source file /src/pkg/container/list/list.go</h1>

  <!-- The Table of Contents is automatically inserted in this <div>.
       Do not delete this <div>. -->
  <div id="nav"></div>

  <!-- Content is HTML-escaped elsewhere -->
  <pre>
<a id="L1"></a><span class="comment">// Copyright 2009 The Go Authors. All rights reserved.</span>
<a id="L2"></a><span class="comment">// Use of this source code is governed by a BSD-style</span>
<a id="L3"></a><span class="comment">// license that can be found in the LICENSE file.</span>

<a id="L5"></a><span class="comment">// The list package implements a doubly linked list.</span>
<a id="L6"></a>package list

<a id="L8"></a><span class="comment">// Element is an element in the linked list.</span>
<a id="L9"></a>type Element struct {
    <a id="L10"></a><span class="comment">// Next and previous pointers in the doubly-linked list of elements.</span>
    <a id="L11"></a><span class="comment">// The front of the list has prev = nil, and the back has next = nil.</span>
    <a id="L12"></a>next, prev *Element;

    <a id="L14"></a><span class="comment">// A unique ID for the list to which this element belongs.</span>
    <a id="L15"></a>id  *byte;

    <a id="L17"></a><span class="comment">// The contents of this list element.</span>
    <a id="L18"></a>Value interface{};
<a id="L19"></a>}

<a id="L21"></a><span class="comment">// Next returns the next list element or nil.</span>
<a id="L22"></a>func (e *Element) Next() *Element { return e.next }

<a id="L24"></a><span class="comment">// Prev returns the previous list element or nil.</span>
<a id="L25"></a>func (e *Element) Prev() *Element { return e.prev }

<a id="L27"></a><span class="comment">// List represents a doubly linked list.</span>
<a id="L28"></a>type List struct {
    <a id="L29"></a>front, back *Element;
    <a id="L30"></a>len         int;
    <a id="L31"></a>id          *byte;
<a id="L32"></a>}

<a id="L34"></a><span class="comment">// Init initializes or clears a List.</span>
<a id="L35"></a>func (l *List) Init() *List {
    <a id="L36"></a>l.front = nil;
    <a id="L37"></a>l.back = nil;
    <a id="L38"></a>l.len = 0;
    <a id="L39"></a>l.id = new(byte);
    <a id="L40"></a>return l;
<a id="L41"></a>}

<a id="L43"></a><span class="comment">// New returns an initialized list.</span>
<a id="L44"></a>func New() *List { return new(List).Init() }

<a id="L46"></a><span class="comment">// Front returns the first element in the list.</span>
<a id="L47"></a>func (l *List) Front() *Element { return l.front }

<a id="L49"></a><span class="comment">// Back returns the last element in the list.</span>
<a id="L50"></a>func (l *List) Back() *Element { return l.back }

<a id="L52"></a><span class="comment">// Remove removes the element from the list.</span>
<a id="L53"></a>func (l *List) Remove(e *Element) {
    <a id="L54"></a>if e.id != l.id {
        <a id="L55"></a>return
    <a id="L56"></a>}
    <a id="L57"></a>if e.prev == nil {
        <a id="L58"></a>l.front = e.next
    <a id="L59"></a>} else {
        <a id="L60"></a>e.prev.next = e.next
    <a id="L61"></a>}
    <a id="L62"></a>if e.next == nil {
        <a id="L63"></a>l.back = e.prev
    <a id="L64"></a>} else {
        <a id="L65"></a>e.next.prev = e.prev
    <a id="L66"></a>}

    <a id="L68"></a>e.prev = nil;
    <a id="L69"></a>e.next = nil;
    <a id="L70"></a>l.len--;
<a id="L71"></a>}

<a id="L73"></a>func (l *List) insertBefore(e *Element, mark *Element) {
    <a id="L74"></a>if mark.prev == nil {
        <a id="L75"></a><span class="comment">// new front of the list</span>
        <a id="L76"></a>l.front = e
    <a id="L77"></a>} else {
        <a id="L78"></a>mark.prev.next = e
    <a id="L79"></a>}
    <a id="L80"></a>e.prev = mark.prev;
    <a id="L81"></a>mark.prev = e;
    <a id="L82"></a>e.next = mark;
    <a id="L83"></a>l.len++;
<a id="L84"></a>}

<a id="L86"></a>func (l *List) insertAfter(e *Element, mark *Element) {
    <a id="L87"></a>if mark.next == nil {
        <a id="L88"></a><span class="comment">// new back of the list</span>
        <a id="L89"></a>l.back = e
    <a id="L90"></a>} else {
        <a id="L91"></a>mark.next.prev = e
    <a id="L92"></a>}
    <a id="L93"></a>e.next = mark.next;
    <a id="L94"></a>mark.next = e;
    <a id="L95"></a>e.prev = mark;
    <a id="L96"></a>l.len++;
<a id="L97"></a>}

<a id="L99"></a>func (l *List) insertFront(e *Element) {
    <a id="L100"></a>if l.front == nil {
        <a id="L101"></a><span class="comment">// empty list</span>
        <a id="L102"></a>l.front, l.back = e, e;
        <a id="L103"></a>e.prev, e.next = nil, nil;
        <a id="L104"></a>l.len = 1;
        <a id="L105"></a>return;
    <a id="L106"></a>}
    <a id="L107"></a>l.insertBefore(e, l.front);
<a id="L108"></a>}

<a id="L110"></a>func (l *List) insertBack(e *Element) {
    <a id="L111"></a>if l.back == nil {
        <a id="L112"></a><span class="comment">// empty list</span>
        <a id="L113"></a>l.front, l.back = e, e;
        <a id="L114"></a>e.prev, e.next = nil, nil;
        <a id="L115"></a>l.len = 1;
        <a id="L116"></a>return;
    <a id="L117"></a>}
    <a id="L118"></a>l.insertAfter(e, l.back);
<a id="L119"></a>}

<a id="L121"></a><span class="comment">// PushFront inserts the value at the front of the list and returns a new Element containing the value.</span>
<a id="L122"></a>func (l *List) PushFront(value interface{}) *Element {
    <a id="L123"></a>if l.id == nil {
        <a id="L124"></a>l.Init()
    <a id="L125"></a>}
    <a id="L126"></a>e := &amp;Element{nil, nil, l.id, value};
    <a id="L127"></a>l.insertFront(e);
    <a id="L128"></a>return e;
<a id="L129"></a>}

<a id="L131"></a><span class="comment">// PushBack inserts the value at the back of the list and returns a new Element containing the value.</span>
<a id="L132"></a>func (l *List) PushBack(value interface{}) *Element {
    <a id="L133"></a>if l.id == nil {
        <a id="L134"></a>l.Init()
    <a id="L135"></a>}
    <a id="L136"></a>e := &amp;Element{nil, nil, l.id, value};
    <a id="L137"></a>l.insertBack(e);
    <a id="L138"></a>return e;
<a id="L139"></a>}

<a id="L141"></a><span class="comment">// InsertBefore inserts the value immediately before mark and returns a new Element containing the value.</span>
<a id="L142"></a>func (l *List) InsertBefore(value interface{}, mark *Element) *Element {
    <a id="L143"></a>if mark.id != l.id {
        <a id="L144"></a>return nil
    <a id="L145"></a>}
    <a id="L146"></a>e := &amp;Element{nil, nil, l.id, value};
    <a id="L147"></a>l.insertBefore(e, mark);
    <a id="L148"></a>return e;
<a id="L149"></a>}

<a id="L151"></a><span class="comment">// InsertAfter inserts the value immediately after mark and returns a new Element containing the value.</span>
<a id="L152"></a>func (l *List) InsertAfter(value interface{}, mark *Element) *Element {
    <a id="L153"></a>if mark.id != l.id {
        <a id="L154"></a>return nil
    <a id="L155"></a>}
    <a id="L156"></a>e := &amp;Element{nil, nil, l.id, value};
    <a id="L157"></a>l.insertAfter(e, mark);
    <a id="L158"></a>return e;
<a id="L159"></a>}

<a id="L161"></a><span class="comment">// MoveToFront moves the element to the front of the list.</span>
<a id="L162"></a>func (l *List) MoveToFront(e *Element) {
    <a id="L163"></a>if e.id != l.id || l.front == e {
        <a id="L164"></a>return
    <a id="L165"></a>}
    <a id="L166"></a>l.Remove(e);
    <a id="L167"></a>l.insertFront(e);
<a id="L168"></a>}

<a id="L170"></a><span class="comment">// MoveToBack moves the element to the back of the list.</span>
<a id="L171"></a>func (l *List) MoveToBack(e *Element) {
    <a id="L172"></a>if e.id != l.id || l.back == e {
        <a id="L173"></a>return
    <a id="L174"></a>}
    <a id="L175"></a>l.Remove(e);
    <a id="L176"></a>l.insertBack(e);
<a id="L177"></a>}

<a id="L179"></a><span class="comment">// Len returns the number of elements in the list.</span>
<a id="L180"></a>func (l *List) Len() int { return l.len }

<a id="L182"></a>func (l *List) iterate(c chan&lt;- interface{}) {
    <a id="L183"></a>for e := l.front; e != nil; e = e.next {
        <a id="L184"></a>c &lt;- e.Value
    <a id="L185"></a>}
    <a id="L186"></a>close(c);
<a id="L187"></a>}

<a id="L189"></a>func (l *List) Iter() &lt;-chan interface{} {
    <a id="L190"></a>c := make(chan interface{});
    <a id="L191"></a>go l.iterate(c);
    <a id="L192"></a>return c;
<a id="L193"></a>}
</pre>

</div>

<div id="footer">
<p>Except as noted, this content is
   licensed under <a href="http://creativecommons.org/licenses/by/3.0/">
   Creative Commons Attribution 3.0</a>.
</div>

<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
var pageTracker = _gat._getTracker("UA-11222381-2");
pageTracker._trackPageview();
</script>
</body>
</html>
<!-- generated at Thu Nov 12 15:42:51 PST 2009 -->

File: go/src/cmd/8l/list.c
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
   "http://www.w3.org/TR/html4/transitional.dtd">
<html>
<head>

  <meta http-equiv="content-type" content="text/html; charset=utf-8">
  <title>Text file src/cmd/8l/list.c</title>

  <link rel="stylesheet" type="text/css" href="../../../doc/style.css">
  <script type="text/javascript" src="../../../doc/godocs.js"></script>

</head>

<body>

  <script>
    // Catch 'enter' key down events and trigger the search form submission.
    function codesearchKeyDown(event) {
      if (event.which == 13) {
        var form = document.getElementById('codesearch');
        var query = document.getElementById('codesearchQuery');
        form.q.value = "lang:go package:go.googlecode.com " + query.value;
        document.getElementById('codesearch').submit();
}      return true;
}
    // Capture the submission event and construct the query parameter.
    function codeSearchSubmit() {
      var query = document.getElementById('codesearchQuery');
      var form = document.getElementById('codesearch');
      form.q.value = "lang:go package:go.googlecode.com " + query.value;
      return true;
}  </script>

<div id="topnav">
  <table summary="">
    <tr>
      <td id="headerImage">
        <a href="../../../index.html"><img src="../../../doc/logo-153x55.png" height="55" width="153" alt="Go Home Page" style="border:0" /></a>
      </td>
      <td>
        <div id="headerDocSetTitle">The Go Programming Language</div>
      </td>
      <td>
        <!-- <table>
          <tr>
            <td>
              <! The input box is outside of the form because we want to add
              a couple of restricts to the query before submitting. If we just
              add the restricts to the text box before submitting, then they
              appear in the box when the user presses 'back'. Thus we use a
              hidden field in the form. However, there's no way to stop the
              non-hidden text box from also submitting a value unless we move
              it outside of the form
              <input type="search" id="codesearchQuery" value="" size="30" onkeydown="return codesearchKeyDown(event);"/>
              <form method="GET" action="http://www.google.com/codesearch" id="codesearch" class="search" onsubmit="return codeSearchSubmit();" style="display:inline;">
                <input type="hidden" name="q" value=""/>
                <input type="submit" value="Code search" />
                <span style="color: red">(TODO: remove for now?)</span>
              </form>
            </td>
          </tr>
          <tr>
            <td>
              <span style="color: gray;">(e.g. &ldquo;pem&rdquo; or &ldquo;xml&rdquo;)</span>
            </td>
          </tr>
        </table> -->
      </td>
    </tr>
  </table>
</div>

<div id="linkList">
  <ul>
    <li class="navhead"><a href="../../../index.html">Home</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Documents</li>
    <li><a href="../../../doc/go_tutorial.html">Tutorial</a></li>
    <li><a href="../../../doc/effective_go.html">Effective Go</a></li>
    <li><a href="../../../doc/go_faq.html">FAQ</a></li>
    <li><a href="../../../doc/go_lang_faq.html">Language Design FAQ</a></li>
    <li><a href="http://www.youtube.com/watch?v=rKnDgT73v8s">Tech talk (1 hour)</a> (<a href="../../../doc/go_talk-20091030.pdf">PDF</a>)</li>
    <li><a href="../../../doc/go_spec.html">Language Specification</a></li>
    <li><a href="../../../doc/go_mem.html">Memory Model</a></li>
    <li><a href="../../../doc/go_for_cpp_programmers.html">Go for C++ Programmers</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">How To</li>
    <li><a href="../../../doc/install.html">Install Go</a></li>
    <li><a href="../../../doc/contribute.html">Contribute code</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Programming</li>
    <li><a href="../../../cmd/index.html">Command documentation</a></li>
    <li><a href="../../../pkg/index.html">Package documentation</a></li>
    <li><a href="../../index.html">Source files</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Help</li>
    <li>#go-nuts on irc.freenode.net</li>
    <li><a href="http://groups.google.com/group/golang-nuts">Go Nuts mailing list</a></li>
    <li><a href="http://code.google.com/p/go/issues/list">Issue tracker</a></li>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Go code search</li>
    <form method="GET" action="http://golang.org/search" class="search">
    <input type="search" name="q" value="" size="25" style="width:80%; max-width:200px" />
    <input type="submit" value="Go" />
    </form>

    <li class="blank">&nbsp;</li>
    <li class="navhead">Last update</li>
	<li>Thu Nov 12 15:48:37 PST 2009</li>
  </ul>
</div>

<div id="content">
  <h1 id="generatedHeader">Text file src/cmd/8l/list.c</h1>

  <!-- The Table of Contents is automatically inserted in this <div>.
       Do not delete this <div>. -->
  <div id="nav"></div>

  <!-- Content is HTML-escaped elsewhere -->
  <pre>
// Inferno utils/8l/list.c
// http://code.google.com/p/inferno-os/source/browse/utils/8l/list.c
//
//	Copyright © 1994-1999 Lucent Technologies Inc.  All rights reserved.
//	Portions Copyright © 1995-1997 C H Forsyth (forsyth@terzarima.net)
//	Portions Copyright © 1997-1999 Vita Nuova Limited
//	Portions Copyright © 2000-2007 Vita Nuova Holdings Limited (www.vitanuova.com)
//	Portions Copyright © 2004,2006 Bruce Ellis
//	Portions Copyright © 2005-2007 C H Forsyth (forsyth@terzarima.net)
//	Revisions Copyright © 2000-2007 Lucent Technologies Inc. and others
//	Portions Copyright © 2009 The Go Authors.  All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the &#34;Software&#34;), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED &#34;AS IS&#34;, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#include	&#34;l.h&#34;
#include	&#34;../ld/lib.h&#34;

void
listinit(void)
{

	fmtinstall(&#39;R&#39;, Rconv);
	fmtinstall(&#39;A&#39;, Aconv);
	fmtinstall(&#39;D&#39;, Dconv);
	fmtinstall(&#39;S&#39;, Sconv);
	fmtinstall(&#39;P&#39;, Pconv);
}

static	Prog	*bigP;

int
Pconv(Fmt *fp)
{
	char str[STRINGSZ];
	Prog *p;

	p = va_arg(fp-&gt;args, Prog*);
	bigP = p;
	switch(p-&gt;as) {
	case ATEXT:
		if(p-&gt;from.scale) {
			sprint(str, &#34;(%ld)	%A	%D,%d,%D&#34;,
				p-&gt;line, p-&gt;as, &amp;p-&gt;from, p-&gt;from.scale, &amp;p-&gt;to);
			break;
		}
	default:
		sprint(str, &#34;(%ld)	%A	%D,%D&#34;,
			p-&gt;line, p-&gt;as, &amp;p-&gt;from, &amp;p-&gt;to);
		break;
	case ADATA:
	case AINIT:
	case ADYNT:
		sprint(str, &#34;(%ld)	%A	%D/%d,%D&#34;,
			p-&gt;line, p-&gt;as, &amp;p-&gt;from, p-&gt;from.scale, &amp;p-&gt;to);
		break;
	}
	bigP = P;
	return fmtstrcpy(fp, str);
}

int
Aconv(Fmt *fp)
{
	int i;

	i = va_arg(fp-&gt;args, int);
	return fmtstrcpy(fp, anames[i]);
}

char*
xsymname(Sym *s)
{
	if(s == nil)
		return &#34;!!noname!!&#34;;
	return s-&gt;name;
}

int
Dconv(Fmt *fp)
{
	char str[40], s[20];
	Adr *a;
	int i;

	a = va_arg(fp-&gt;args, Adr*);
	i = a-&gt;type;
	if(i &gt;= D_INDIR &amp;&amp; i &lt; 2*D_INDIR) {
		if(a-&gt;offset)
			sprint(str, &#34;%ld(%R)&#34;, a-&gt;offset, i-D_INDIR);
		else
			sprint(str, &#34;(%R)&#34;, i-D_INDIR);
		goto brk;
	}
	switch(i) {

	default:
		sprint(str, &#34;%R&#34;, i);
		break;

	case D_NONE:
		str[0] = 0;
		break;

	case D_BRANCH:
		if(bigP != P &amp;&amp; bigP-&gt;pcond != P)
			if(a-&gt;sym != S)
				sprint(str, &#34;%lux+%s&#34;, bigP-&gt;pcond-&gt;pc,
					a-&gt;sym-&gt;name);
			else
				sprint(str, &#34;%lux&#34;, bigP-&gt;pcond-&gt;pc);
		else
			sprint(str, &#34;%ld(PC)&#34;, a-&gt;offset);
		break;

	case D_EXTERN:
		sprint(str, &#34;%s+%ld(SB)&#34;, xsymname(a-&gt;sym), a-&gt;offset);
		break;

	case D_STATIC:
		sprint(str, &#34;%s&lt;%d&gt;+%ld(SB)&#34;, xsymname(a-&gt;sym),
			a-&gt;sym-&gt;version, a-&gt;offset);
		break;

	case D_AUTO:
		sprint(str, &#34;%s+%ld(SP)&#34;, xsymname(a-&gt;sym), a-&gt;offset);
		break;

	case D_PARAM:
		if(a-&gt;sym)
			sprint(str, &#34;%s+%ld(FP)&#34;, a-&gt;sym-&gt;name, a-&gt;offset);
		else
			sprint(str, &#34;%ld(FP)&#34;, a-&gt;offset);
		break;

	case D_CONST:
		sprint(str, &#34;$%ld&#34;, a-&gt;offset);
		break;

	case D_CONST2:
		sprint(str, &#34;$%ld-%ld&#34;, a-&gt;offset, a-&gt;offset2);
		break;

	case D_FCONST:
		sprint(str, &#34;$(%.8lux,%.8lux)&#34;, a-&gt;ieee.h, a-&gt;ieee.l);
		break;

	case D_SCONST:
		sprint(str, &#34;$\&#34;%S\&#34;&#34;, a-&gt;scon);
		break;

	case D_ADDR:
		a-&gt;type = a-&gt;index;
		a-&gt;index = D_NONE;
		sprint(str, &#34;$%D&#34;, a);
		a-&gt;index = a-&gt;type;
		a-&gt;type = D_ADDR;
		goto conv;
	}
brk:
	if(a-&gt;index != D_NONE) {
		sprint(s, &#34;(%R*%d)&#34;, a-&gt;index, a-&gt;scale);
		strcat(str, s);
	}
conv:
	fmtstrcpy(fp, str);
//	if(a-&gt;gotype)
//		fmtprint(fp, &#34;«%s»&#34;, a-&gt;gotype-&gt;name);
	return 0;
}

char*	regstr[] =
{
	&#34;AL&#34;,		/* [D_AL] */
	&#34;CL&#34;,
	&#34;DL&#34;,
	&#34;BL&#34;,
	&#34;AH&#34;,
	&#34;CH&#34;,
	&#34;DH&#34;,
	&#34;BH&#34;,

	&#34;AX&#34;,		/* [D_AX] */
	&#34;CX&#34;,
	&#34;DX&#34;,
	&#34;BX&#34;,
	&#34;SP&#34;,
	&#34;BP&#34;,
	&#34;SI&#34;,
	&#34;DI&#34;,

	&#34;F0&#34;,		/* [D_F0] */
	&#34;F1&#34;,
	&#34;F2&#34;,
	&#34;F3&#34;,
	&#34;F4&#34;,
	&#34;F5&#34;,
	&#34;F6&#34;,
	&#34;F7&#34;,

	&#34;CS&#34;,		/* [D_CS] */
	&#34;SS&#34;,
	&#34;DS&#34;,
	&#34;ES&#34;,
	&#34;FS&#34;,
	&#34;GS&#34;,

	&#34;GDTR&#34;,		/* [D_GDTR] */
	&#34;IDTR&#34;,		/* [D_IDTR] */
	&#34;LDTR&#34;,		/* [D_LDTR] */
	&#34;MSW&#34;,		/* [D_MSW] */
	&#34;TASK&#34;,		/* [D_TASK] */

	&#34;CR0&#34;,		/* [D_CR] */
	&#34;CR1&#34;,
	&#34;CR2&#34;,
	&#34;CR3&#34;,
	&#34;CR4&#34;,
	&#34;CR5&#34;,
	&#34;CR6&#34;,
	&#34;CR7&#34;,

	&#34;DR0&#34;,		/* [D_DR] */
	&#34;DR1&#34;,
	&#34;DR2&#34;,
	&#34;DR3&#34;,
	&#34;DR4&#34;,
	&#34;DR5&#34;,
	&#34;DR6&#34;,
	&#34;DR7&#34;,

	&#34;TR0&#34;,		/* [D_TR] */
	&#34;TR1&#34;,
	&#34;TR2&#34;,
	&#34;TR3&#34;,
	&#34;TR4&#34;,
	&#34;TR5&#34;,
	&#34;TR6&#34;,
	&#34;TR7&#34;,

	&#34;NONE&#34;,		/* [D_NONE] */
};

int
Rconv(Fmt *fp)
{
	char str[20];
	int r;

	r = va_arg(fp-&gt;args, int);
	if(r &gt;= D_AL &amp;&amp; r &lt;= D_NONE)
		sprint(str, &#34;%s&#34;, regstr[r-D_AL]);
	else
		sprint(str, &#34;gok(%d)&#34;, r);

	return fmtstrcpy(fp, str);
}

int
Sconv(Fmt *fp)
{
	int i, c;
	char str[30], *p, *a;

	a = va_arg(fp-&gt;args, char*);
	p = str;
	for(i=0; i&lt;sizeof(double); i++) {
		c = a[i] &amp; 0xff;
		if(c &gt;= &#39;a&#39; &amp;&amp; c &lt;= &#39;z&#39; ||
		   c &gt;= &#39;A&#39; &amp;&amp; c &lt;= &#39;Z&#39; ||
		   c &gt;= &#39;0&#39; &amp;&amp; c &lt;= &#39;9&#39;) {
			*p++ = c;
			continue;
		}
		*p++ = &#39;\\&#39;;
		switch(c) {
		default:
			if(c &lt; 040 || c &gt;= 0177)
				break;	/* not portable */
			p[-1] = c;
			continue;
		case 0:
			*p++ = &#39;z&#39;;
			continue;
		case &#39;\\&#39;:
		case &#39;&#34;&#39;:
			*p++ = c;
			continue;
		case &#39;\n&#39;:
			*p++ = &#39;n&#39;;
			continue;
		case &#39;\t&#39;:
			*p++ = &#39;t&#39;;
			continue;
		}
		*p++ = (c&gt;&gt;6) + &#39;0&#39;;
		*p++ = ((c&gt;&gt;3) &amp; 7) + &#39;0&#39;;
		*p++ = (c &amp; 7) + &#39;0&#39;;
	}
	*p = 0;
	return fmtstrcpy(fp, str);
}

void
diag(char *fmt, ...)
{
	char buf[STRINGSZ], *tn;
	va_list arg;

	tn = &#34;??none??&#34;;
	if(curtext != P &amp;&amp; curtext-&gt;from.sym != S)
		tn = curtext-&gt;from.sym-&gt;name;
	va_start(arg, fmt);
	vseprint(buf, buf+sizeof(buf), fmt, arg);
	va_end(arg);
	print(&#34;%s: %s\n&#34;, tn, buf);

	nerrors++;
	if(nerrors &gt; 20) {
		print(&#34;too many errors\n&#34;);
		errorexit();
	}
}
</pre>

</div>

<div id="footer">
<p>Except as noted, this content is
   licensed under <a href="http://creativecommons.org/licenses/by/3.0/">
   Creative Commons Attribution 3.0</a>.
</div>

<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
var pageTracker = _gat._getTracker("UA-11222381-2");
pageTracker._trackPageview();
</script>
</body>
</html>
<!-- generated at Thu Nov 12 15:42:51 PST 2009 -->
